### ISSUE TITLE
Add iOS 18 compatible widget

### ISSUE THREAD

**Issue by victorlin** [OneBusAway/onebusaway-ios#748]
iOS 18 [removed support](https://9to5mac.com/2024/07/23/psa-ios-18-is-going-to-disable-your-old-today-view-widgets/) for the Today View widget added in #32. Would it be possible to port the same or similar functionality to an iOS 18 compatible widget?

**Comment by aaronbrethorst** [comment]
Hi Victor - Thanks for opening this issue! At least in the iOS 17 timeframe, which was the last time I checked, Apple limited WidgetKit widgets to updating only a handful of times per hour—and on a schedule completely out of the hands of developers and users. Do you happen to know if there have been any changes on this front?

If you have any experience with iOS, we'd love your help in building a new WidgetKit-based experience, but either way I hope we can ship an update including this soon.

**Comment by victorlin** [comment]
Unfortunately I don't have any iOS experience, but I actively used the Today View widget and was sad to see it disappear after upgrading to iOS 18. Glad to hear there is interest in bringing it back.

A quick internet search makes me think it may be possible with [app intents](https://developer.apple.com/documentation/widgetkit/making-a-configurable-widget). From [doc page on widget updates](https://developer.apple.com/documentation/widgetkit/keeping-a-widget-up-to-date):

> WidgetKit doesn’t count reloads against your widget’s budget include when ... The widget performs an app intent, such as when the user taps a button or toggles a switch.

Examples:

- [How to add a reload button to an iOS Widget? - Stack Overflow](https://stackoverflow.com/a/76731458)
- [iOS 17 WidgetKit animations (using a refresh button) : r/SwiftUI](https://www.reddit.com/r/SwiftUI/comments/17ro7st/)

**Comment by manu-r12** [comment]
Can i try working on this one? 
I am familiar with widget kit :)

**Comment by aaronbrethorst** [comment]
@manu-r12 sure, please do!

**Comment by manu-r12** [comment]
<img width="1037" alt="Screenshot 2024-10-16 at 11 44 42 PM" src="https://github.com/user-attachments/assets/7e158c64-9807-4200-b79a-62a2005996ba">
I'm halfway through building the widget and just wanted to check in. Should I keep the current UI as it is, or would you like me to make any changes?

**Comment by aaronbrethorst** [comment]
@manu-r12 this is looking great! The one thing I'd like to see change is to add a bit of spacing in between each arrival-departure and its neighbors:

```
Name                     11m
arrives 1 min late  23m
                               49m

Name 2                  11m
arrives 1 min late  23m
                               49m
```

**Comment by manu-r12** [comment]
## This is almost done look, let me know if there's anything you want me to change. 😀

https://github.com/user-attachments/assets/a5fa8555-804a-4e62-b100-31c2ca543b45

https://github.com/user-attachments/assets/86e8a0c0-f5f4-4c9c-a7ca-1296d935c554



There are a few small issues though. Just like with the old Today View extension, WidgetKit widgets can't resize themselves based on the content.

but there is a workaround for this which is very simple , for example, in the medium widget size, users will only be able to view two bookmarks at a time. To address this, users can choose which bookmarks to display by reordering them in the bookmark view. The top two bookmarks will appear on the widget. The same goes for the large widget size, which can display up to six bookmarks

### I'm currently working on:
- Another solution for this issue.
- A more efficient way to fetch the bookmark data since widgets are lightweight.

#### just let me know if there's anything.... 😊

**Comment by aaronbrethorst** [comment]
I think that's looking great, @manu-r12!

**Comment by manu-r12** [comment]
should i make a PR?

**Comment by aaronbrethorst** [comment]
yes please!

**Comment by aaronbrethorst** [comment]
Fixed in #753 - thank you @manu-r12!
